readme.md
# Terms & Conditions Connector of IT-Recht Kanzlei

Transfers the legal texts of “IT-Recht Kanzlei München” via API into your WordPress-installation and keeps your judicial pages up to date.

![https://www.it-recht-kanzlei.de/agb-starterpaket.php](https://ps.w.org/agb-connector/assets/banner-772x250.png?rev=2167983)

## Description

Due to the **Terms & Conditions Connector of IT-Recht Kanzlei**, the legal texts needn’t be transferred from the clients’ portal to your website by yourself.

Create the judicial pages as empty pages in WordPress and set up the connection to the clients’ portal of IT-Recht Kanzlei. From now on, your judicial pages are always up to date, as amendments of the legal texts in the clients’ portal are immediately transferred to WordPress.

To use this plugin, you need an access to the clients’ portal of “IT-Recht Kanzlei München”. You are no client yet? On the website of “[IT-Recht Kanzlei München](https://www.it-recht-kanzlei.de/agb-starterpaket.php)”, you can find all protection packages.

### Features

* Automatic transfer of legal texts from “IT-Recht Kanzlei München” to your WordPress-installation
* Automatic update of legal texts in WordPress in case of amendments in the clients’ portal
* Suitable for both WordPress websites/blogs and WordPress shops, irrespective of your shop system
* Appends selected legal texts as PDF to the WooCommerce-Email “New Order”
* Simple connection to the clients’ portal of “IT-Recht Kanzlei München” via API-Token and URL
* Use the shortcodes [agb_terms], [agb_privacy], [agb_revocation], [agb_imprint] to output the legal text e.g. in a page builder element.
* Supports foreign / foreign language legal texts

### Benefits

* Always up-to-date and dunning-proof legal texts for your WordPress
* Automatic updates via Terms & Conditions Connector
* Self-evident: legal liability (As other legal offices, “IT-Recht Kanzlei” bears liability for the dunning-proof legal texts in the context of legal provisions in force)
* Full flexibility: As necessary, the terms and conditions Connector is cancelable at the end of a month any time
* Fair: monthly payment (No annual fee ahead)

### Support 

You can find technical support for this plugin in the [wordpress.org forum](https://wordpress.org/support/plugin/agb-connector).

Please read the FAQ (frequently asked questions) (https://wordpress.org/support/plugin/agb-connector/faq) first and make sure you have installed the newest version of the plugin before contacting us.

If you have questions concerning the legal texts, please contact “IT-Recht Kanzlei München” directly via +49 89 13014330 or info@it-recht-kanzlei.de.

**Made by [Inpsyde](https://inpsyde.com) &middot; We love WordPress**

## Installation

### Minimum requirements
* WordPress version >= 5.0
* PHP Version >= 7.2
* Access to a protection package of “IT-Recht Kanzlei München”

### Installation via WordPress Backend

In the WordPress Backend, go to *Plugins => Add New*. Search for the **Terms & Conditions Connector of IT-Recht Kanzlei**. Click on *Install Now* and then *Activate*.


### Setting the automatic transfer of legal texts

To set the transfer of the legal texts, go to *Settings => IT-Recht Kanzlei*. 

1. Generate the API-Token by clicking “Regenerate”.
2. Assign a page of your WordPress-installation to each legal text. Eventually, you need to create these pages first.
3. Save your settings.
4. First, the legal texts need to be configured in the clients’ portal of “IT-Recht Kanzlei München”. Then, you set up the data interface by indicating the URL and the API-Token. Once after the installation, you need to transfer the texts manually. After that, amendments of the legal texts are transferred automatically. (see screenshots.)


### Appending the legal texts to WooCommerce-Emails 

As far as you run a WordPress shop based on WooCommerce, you have the possibility to append selected legal texts as PDF to the Email “New Order”. Therefore, you simply need to go to *Settings => IT-Recht Kanzlei* and then place a hook at your selected legal texts at “Send PDF with WooCommerce order on hold email.”



## Frequently Asked Questions

Here, you find answers to frequently asked questions about this plugin.

### Do I need an access to “IT-Recht Kanzlei München” to use the plugin Terms & Conditions Connector of IT-Recht Kanzlei?

Yes, you need to have booked one of the [protection packages of “IT-Recht Kanzlei München”](https://www.it-recht-kanzlei.de/agb-starterpaket.php) to transfer legal texts to your WordPress-installation.

### Can I use the plugin "Terms & Conditions Connector of IT-Recht Kanzlei" both for websites/blogs and shops?

Yes. The plugin does not distinguish shops or “simple” websites. The only thing you need to do is assigning a page in your WordPress-Installation to every single legal text being made by IT-Recht Kanzlei.

### Concerning Online-Shops: Does the shop need to be based on WooCommerce?

No. The plugin does not require special demands for your WordPress shop. Merely, there must be the appropriate pages to transfer the legal texts.

## Screenshots
![Terms & Conditions Connector of IT-Recht Kanzlei - Screenshot1](https://ps.w.org/agb-connector/assets/screenshot-1.png?rev=2169425)
<p style="text-align: center;">Terms & Conditions Connector of IT-Recht Kanzlei – Settings Page
</p>

![Terms & Conditions Connector of IT-Recht Kanzlei - Screenshot2](https://ps.w.org/agb-connector/assets/screenshot-2.png?rev=2167983)
<p style="text-align: center;">Protection package of IT-Recht Kanzlei München – Registration to the clients’ portal.
</p>


![Terms & Conditions Connector of IT-Recht Kanzlei - Screenshot5](https://ps.w.org/agb-connector/assets/screenshot-5.png?rev=2167983)
<p style="text-align: center;">Protection package of IT-Recht Kanzlei München – Overview of legal texts
</p>

![Terms & Conditions Connector of IT-Recht Kanzlei - Screenshot6](https://ps.w.org/agb-connector/assets/screenshot-6.png?rev=2167983)
<p style="text-align: center;">Protection package of IT-Recht Kanzlei München – Configuration of legal texts</p>

![Terms & Conditions Connector of IT-Recht Kanzlei - Screenshot3](https://ps.w.org/agb-connector/assets/screenshot-3.png?rev=2167983)
<p style="text-align: center;">Protection package of IT-Recht Kanzlei München – Choose shop system</p>

![Terms & Conditions Connector of IT-Recht Kanzlei - Screenshot4](https://ps.w.org/agb-connector/assets/screenshot-4.png?rev=2167983)
<p style="text-align: center;">Protection package of IT-Recht Kanzlei München – API-Token and Shop URL</p>





changelog.md
# Changelog

## Version 3.0.0
* Added: Hide title option
* Changed: Documents now saved as `wp_block` post type
* Changed: All incoming documents saved, no preconfigured allocation needed.
* Changed: Updated settings page appearance.
* Changed: Document can be saved if PDF required but not attached.

## Version 2.0.2
* Fixed: Fixed displaying of Shop URL when WPML is used
* Fixed: Error on reset password page because of type checking

## Version 2.0.1

* Fixed: WPML Support: translated pages can't be selected
* Fixed: Saved Page ID when no PAge is selected
* Fixed: Error message in API when legal page is not configured for text type
* Fixed: Adding new site will not work when special chars in site title

## Version 2.0.0

* Added: Support for multilanguage legal texts
* Added: Legal text page title will now be updated
* Added: Transmitting of legal page url
* Changed: PHP 5.4 is now required
* Changed: Append PDFs also to WooCommerce on hold emails
* Changed: Get PDF file name from the API

## Version 1.0.4

* Added: Shortcode for legal text

## Version 1.0.3

* Improved: Use WordPress internal file get functions
* Improved: Hide DOC ID
* Fixed: Where you couldn't regenerate your API-Token

## Version 1.0.2

* Fixed: Plugindirectory

## Version 1.0.1

* Fixed: Plugindirectory

## Version 1.0

* Init Release


src/Settings.php
<?php

declare(strict_types=1);

namespace Inpsyde\AGBConnector;

use Inpsyde\AGBConnector\CustomExceptions\GeneralException;
use Inpsyde\AGBConnector\Document\DocumentInterface;
use Inpsyde\AGBConnector\Document\DocumentPageFinder\DocumentFinderInterface;
use Inpsyde\AGBConnector\Document\Map\WpPostMetaFields;
use Inpsyde\AGBConnector\Document\Repository\DocumentRepository;
use Inpsyde\AGBConnector\Settings\DocumentsTable;
use InvalidArgumentException;

/**
 * Class Settings
 */
class Settings
{

    const AJAX_ACTION = 'agb-update-document-settings';

    const MIGRATION_FAILED_FLAG_OPTION_NAME = 'agbc_flag_migration_failed';

    /**
     * @var DocumentRepository
     */
    protected $repository;

    /**
     * @var DocumentFinderInterface
     */
    protected $documentPageFinder;
    /**
     * @var ShortCodes
     */
    protected $shortCodes;

    /**
     * @var string
     */
    protected $menuPageSlug;

    /**
     * The save message.
     *
     * @var string
     */
    private $message = '';

    /**
     * Message to be displayed if migration failed.
     *
     * @var string
     */
    protected $migrationFailedMessage = '';

    /**
     * @param DocumentRepository $repository
     * @param DocumentFinderInterface $documentPageFinder
     * @param ShortCodes $shortCodes
     */
    public function __construct(
        DocumentRepository $repository,
        DocumentFinderInterface $documentPageFinder,
        ShortCodes $shortCodes
    ) {

        $this->repository = $repository;
        $this->documentPageFinder = $documentPageFinder;
        $this->migrationFailedMessage = sprintf(
            __(
                'AGB Connector couldn\'t migrate your documents from the old plugin version. Please, contact %1$sInpsyde support%2$s for help',
                'agb-connector'
            ),
            '<a href="mailto:agb-connector@inpsyde.com" target="_blank">',
            '</a>'
        );
        $this->shortCodes = $shortCodes;
        $this->menuPageSlug = 'agb_connector_settings';
    }

    public function init()
    {
        add_action('wp_ajax_' . self::AJAX_ACTION, [$this, 'handleAjaxRequest']);
        add_action('before_delete_post', [$this, 'filterRedirectAfterDocumentDeleted']);
    }

    /**
     * Update document settings on AJAX request.
     */
    public function handleAjaxRequest(): void
    {
        check_admin_referer(self::AJAX_ACTION, 'nonce');

        $documentId = (int) filter_input(INPUT_POST, 'documentId', FILTER_SANITIZE_NUMBER_INT);
        $document = $this->repository->getDocumentById($documentId);

        if ($document === null) {
            wp_send_json_error(
                [
                    'message' => __('Document not found.', 'agb-connector'),
                ]
            );
        }

        $fieldName = filter_input(INPUT_POST, 'fieldName', FILTER_SANITIZE_STRING);
        $fieldValue = filter_input(INPUT_POST, 'fieldValue', FILTER_VALIDATE_BOOLEAN);

        try {
            $this->updateDocumentSettings($document, $fieldName, $fieldValue);
        } catch (InvalidArgumentException $exception) {
            wp_send_json_error(['message' => $exception->getMessage()]);
        }

        wp_send_json_success(['nonce' => wp_create_nonce(self::AJAX_ACTION)]);
    }

    /**
     * Add the menu entry
     */
    public function addMenu()
    {
        $hook = add_options_page(
            __('Terms & Conditions Connector of IT-Recht Kanzlei', 'agb-connector'),
            'AGB Connector',
            'edit_pages',
            $this->menuPageSlug,
            [
                $this,
                'page',
            ]
        );

        if ($hook !== false) {
            add_action('load-' . $hook, [$this, 'load']);
        }
    }

    /**
     * Add settings link to plugin actions.
     *
     * @param array $links The links.
     *
     * @return array
     */
    public function addActionLinks($links)
    {
        $addLinks = [
            '<a href="' . admin_url('options-general.php?page=agb_connector_settings') . '">' . __(
                'Settings',
                'agb-connector'
            ) . '</a>',
        ];

        return array_merge($addLinks, $links);
    }

    /**
     * All things must done in load
     *
     * phpcs:disable Generic.Metrics.NestingLevel.TooHigh
     * phpcs:disable Inpsyde.CodeQuality.FunctionLength.TooLong
     */
    public function load()
    {
        wp_enqueue_style(
            'agb-connector',
            plugins_url('/assets/css/style.css', __DIR__),
            [],
            Plugin::VERSION,
            'all'
        );

        wp_enqueue_script(
            'agb-connector',
            plugins_url('/assets/js/settings.js', __DIR__),
            ['jquery'],
            Plugin::VERSION,
            true
        );

        wp_localize_script('agb-connector', 'agbConnectorSettings', [
            'action' => self::AJAX_ACTION,
            'nonce' => wp_create_nonce(self::AJAX_ACTION),
            ]);

        $getRegen = filter_input(INPUT_GET, 'regen', FILTER_SANITIZE_NUMBER_INT);
        if (null !== $getRegen) {
            check_admin_referer('agb-connector-settings-page-regen');
            $userAuthToken = md5(wp_generate_password(32, true, true));
            update_option(Plugin::OPTION_USER_AUTH_TOKEN, $userAuthToken);
            $this->message = __('New APT-Token generated.', 'agb-connector');
        }
    }

    /**
     * The settings page content
     *
     * phpcs:disable Inpsyde.CodeQuality.FunctionLength.TooLong
     */
    public function page()
    {
        $table = new DocumentsTable(
            $this->repository,
            $this->documentPageFinder,
            $this->shortCodes,
            [
                'singular' => __('Document', 'agb-connector'),
                'plural' => __('Documents', 'agb-connector'),
                'ajax' => true,
            ]
        );

        ?>
        <div class="wrap" id="agb-connector-settings">
            <h2>
                <?php
                printf(
                    '%s &rsaquo; %s',
                    esc_html__('Settings', 'agb-connector'),
                    esc_html__('Terms & Conditions Connector of IT-Recht Kanzlei', 'agb-connector')
                );
                ?>
            </h2>
            <div class="box-container">
                <div id="it-recht-kanzlei" class="metabox-holder postbox">
                    <div class="inside">
                        <a href="https://www.it-recht-kanzlei.de" class="it-kanzlei-logo"
                           title="IT-Recht Kanzlei München">IT-Recht Kanzlei München</a><br>
                    </div>
                </div>
                <div id="inpsyde" class="metabox-holder postbox">
                    <div class="inside">
                        <a href="
                        <?php esc_html_e(
                            'https://inpsyde.com/en/?utm_source=Plugin&utm_medium=Banner&utm_campaign=Inpsyde',
                            'agb-connector'
                        ); ?>
                        " class="inpsyde-logo" title="
                        <?php esc_html_e('An Inpsyde GmbH Product', 'agb-connector'); ?>
                        ">Inpsyde GmbH</a>
                        <br>
                    </div>
                </div>
                <div id="inpsyde" class="metabox-holder postbox">
                    <div class="inside">
                        <h3>Support</h3>
                        <p>
                            <?php esc_html_e(
                                'If you have questions please contact “IT-Recht Kanzlei München” directly',
                                'agb-connector'
                            ); ?>
                            <br/>
                            <?php esc_html_e('via +49 89 13014330 or ', 'agb-connector'); ?>
                            <a href="mailto:info@it-recht-kanzlei.de">info@it-recht-kanzlei.de</a>
                        </p>
                    </div>
                </div>
            </div>
            <div class="settings-container">
                <?php
                if ($this->message) {
                    echo '<div id="message" class="updated"><p>' . esc_html($this->message) . '</p></div>';
                }
                if (get_option(self::MIGRATION_FAILED_FLAG_OPTION_NAME, false)) {
                    echo  '<div id="agbc-migration-failed" class="notice notice-warning"><p>' . wp_kses_post($this->migrationFailedMessage) . '</p></div>';
                }
                ?>
                    <table class="form-table">
                        <tr valign="top">
                            <th scope="row">
                                <label for="regen">
                                    <?php esc_html_e('API-Token', 'agb-connector'); ?>
                                </label>
                            </th>
                            <td>
                                <p>
                                    <code><?php echo esc_attr(get_option(Plugin::OPTION_USER_AUTH_TOKEN)); ?></code>
                                    <a class="button" href="
                                    <?php echo esc_url(
                                        wp_nonce_url(
                                            add_query_arg(
                                                [
                                                    'regen' => '',
                                                    'page' => 'agb_connector_settings',
                                                ],
                                                admin_url('options-general.php')
                                            ),
                                            'agb-connector-settings-page-regen'
                                        )
                                    );
                                    ?>">
                                        <?php esc_html_e('Regenerate', 'agb-connector'); ?>
                                    </a>
                                </p>
                                <p class="description">
                                    <?php esc_html_e(
                                        'If you change the token, this must also be adjusted in the client portal.',
                                        'agb-connector'
                                    ); ?>
                                </p>
                            </td>
                        </tr>
                        <tr valign="top">
                            <th scope="row">
                                <label for="regen">
                                    <?php esc_html_e('Your shop URL', 'agb-connector'); ?>
                                </label>
                            </th>
                            <td>
                                <p>
                                    <code>
                                        <?php
                                            //Directly use option that that WPML can't change it
                                            $homeUrl = trailingslashit(get_option('home'));
                                            $homeUrl = set_url_scheme($homeUrl);
                                            echo esc_attr($homeUrl);
                                        ?>
                                    </code>
                                </p>
                            </td>
                        </tr>
                    </table>
            </div>
            <?php
                $table->prepare_items();

                echo '<form method="post">';
                $table->display();
                echo '</form>';
            ?>
        </div>
        <?php
    }

    /**
     * Update document settings.
     *
     * @param DocumentInterface $document
     * @param string $fieldName
     * @param bool $fieldValue
     *
     * @throws InvalidArgumentException If document settings field not found.
     * @throws GeneralException
     */
    protected function updateDocumentSettings(DocumentInterface $document, string $fieldName, bool $fieldValue): void
    {
        switch ($fieldName) {
            case 'store_pdf':
                $document->getSettings()->setSavePdf($fieldValue);
                if (! $fieldValue) {
                    $document->getSettings()->setAttachToWcEmail(false);
                }
                break;
            case 'attach_pdf_to_wc':
                $document->getSettings()->setAttachToWcEmail($fieldValue);
                break;
            case 'hide_title':
                $document->getSettings()->setHideTitle($fieldValue);
                break;
            default:
                throw new InvalidArgumentException(
                    __('Failed to update document: no such field found.', 'agb-connector')
                );
        }

        $this->repository->saveDocument($document);
    }

    /**
     * Filter redirect url after document deletion.
     *
     * @param int $postId Deleted document id
     */
    public function filterRedirectAfterDocumentDeleted($postId): void
    {
        $isDocument = get_post_meta($postId, WpPostMetaFields::WP_POST_DOCUMENT_TYPE);

        if (! $isDocument) {
            return;
        }

        add_filter('wp_redirect', function () {
            return menu_page_url($this->menuPageSlug, false);
        });
    }
}
